design-docs/improved-console.md

```diff
@@ -0,0 +1,422 @@
+* Progress bar is truncated at width of Console
+* Fuzz testing to protect against any security issues arising from using user input
+
+### Open issues
+* Could this concept allow customization of ProgressLine format, providing even greater flexibility?
+
+## Story: Gradle-provided slick themes
+Configure the default `org.gradle.console.progressbar.*` prefix and suffix based on a single 
+property.
+
+For example, a [powerline](https://github.com/powerline/powerline) theme, we would change the
+default progress bar properties to:
+
+```bash
+org.gradle.console.progressbar.succeeded.prefix="[42m"   #BG_GREEN
+org.gradle.console.progressbar.failed.prefix="[41m"      #BG_RED
+org.gradle.console.progressbar.inprogress.prefix="[43m"  #BG_YELLOW
+org.gradle.console.progressbar.unstarted.prefix="[40m"   #BG_BLACK
+org.gradle.console.progressbar.region.suffix=""
+org.gradle.console.progressbar.completechar=" "
+```
+
+### Implementation
+A `PowerlineProgressBarRegion` would extend `ProgressBarRegion` with the defaults
+given above. `BuildProgressBarListener` would detect the use of a Gradle property
+`org.gradle.console.theme=powerline` and substitute `PowerlineProgressBarRegion` 
+instead of the `DefaultProgressBarRegion`.
+
+## Story: (Optional) Customizable StatusBarFormat
+Allow users to configure the format of `StatusLine` entries in the status TextArea.
+
+Aspects that can be customized:
+ * `StatusBarLine` Prefix
+ * `StatusBarLine` Suffix
+
+### Implementation
+Introduce Gradle properties that allow this customization (with defaults):
+```bash
+org.gradle.console.statusline.prefix=" > "
+org.gradle.console.statusline.suffix=""
+```
+
+### Test Cases
+* Fuzz testing ensures no other parts of build are affected by user input
+
+### Open issues
+* How is intra-operation progress ([55 / 1234] or spinner) affected?
+
+## Story: (Optional) Indicate Gradle's continued progress in absence of intra-operation updates
+Provide motion in output to indicate Gradle's working if build operations are long-running.
+
+Options:
+* Incrementing time
+* Blinking a small area of progress bar
+* Blinking indicator on progress operations
+* Adjusting color of progress operations (may require 256 colors)
+* Text-based spinner (e.g. "┘└┌┐" or "|/-\\")
+
+### Implementation
+`Operations` data store knows the time since each operation was started. The 
+`StatusBarFormatter` can choose to render an operation slightly differently depending
+on its lifespan.
+
+### Open issues
+* Best choice to render from given options above
+
+## Story: (Optional) Detect terminals that support UTF-8 and provide richer UI elements by default
+Support detection and use of UCS-2 (default supported by Windows) or UTF-8 characters could further polish this if we choose.
+
+This would allow us to use characters like ᐅ or ► supported by a majority of monospace fonts.
+More granular width for progress bars: ' ', '▏', '▎', '▎', '▍', '▍', '▌', '▌', '▋', '▋', '▊', '▊', '▉', '▉', '█'
```

Comment:
Let's see what we can come up with. I think it's going to be really tough to get one that looks great for 80% of users. I want something that is viral because it's so awesome — that's likely not achievable.